File: ios/PanoRtc/React/PromiseCallback.swift
//
//  PromiseCallback.swift
//  pano_rtc
//
//  Copyright © 2021 Pano. All rights reserved.
//

import Foundation
import PanoRtc

@objc(PromiseCallback)
class PromiseCallback: NSObject, Callback {
    private var resolve: RCTPromiseResolveBlock?
    private var reject: RCTPromiseRejectBlock?

    init(_ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        self.resolve = resolve
        self.reject = reject
    }

    func success(_ data: Any?) {
        resolve?(data)
    }

    func failure(_ code: String) {
        reject?(code, nil, nil)
    }
}

File: android/src/main/java/video/pano/rtc/react/PromiseCallback.kt
package video.pano.rtc.react

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.pano.rtc.api.Constants
import video.pano.rtc.native.Callback

class PromiseCallback(private val promise: Promise?) : Callback() {
    override fun success(data: Any?) {
        when (data) {
            null, is Boolean, is Int, is String -> promise?.resolve(data)
            is Constants.QResult -> promise?.resolve(data.value)
            is Number -> promise?.resolve(data.toDouble())
            is Array<*> -> promise?.resolve(Arguments.makeNativeArray<Any>(arrayOf(data)))
            is List<*> -> promise?.resolve(Arguments.makeNativeArray(data))
            is Map<*, *> -> {
                val map = mutableMapOf<String, Any?>()
                data.forEach {
                    if (it.key is String) {
                        map[it.key as String] = it.value
                    }
                }
                promise?.resolve(Arguments.makeNativeMap(map))
            }
            else -> promise?.reject(IllegalArgumentException("Could not convert " + data.javaClass))
        }
    }

    override fun failure(code: String, message: String) {
        promise?.reject(code, message)
    }
}
